apis/v1alpha1/httproute_types.go

```diff
@@ -212,32 +227,70 @@ type HTTPRouteMatch struct {
+)
+
+// HTTPRouteFilter defines additional processing steps that must be completed
+// during the request or response lifecycle.
+// HTTPRouteFilters are meant as an extension point to express additional
+// processing that may be done in Gateway implementations. Some examples include
+// request or response modification, implementing authentication strategies,
+// rate-limiting, and traffic shaping.
+// API guarantee/conformance is defined based on the type of the filter.
+// TODO(hbagdi): re-render CRDs once controller-tools supports union tags:
+// - https://github.com/kubernetes-sigs/controller-tools/pull/298
+// - https://github.com/kubernetes-sigs/controller-tools/issues/461
+// +union
+type HTTPRouteFilter struct {
+	// Type identifies the filter to execute.
+	// Types are classified into three conformance-levels (similar to
+	// other locations in this API):
+	// - Core and extended: These filter types and their corresponding configuration
+	//   is defined in this package. All implementations must implement
+	//   the core filters. Implementers are encouraged to support extended filters.
+	//   Definitions for filter-specific configuration for these
+	//   filters is defined in this package.
+	// - Custom: These filters are defined and supported by specific vendors.
+	//   In the future, filters showing convergence in behavior across multiple
+	//   implementations will be considered for inclusion in extended or core
+	//   conformance rings. Filter-specific configuration for such filters
+	//   is specified using the ExtensionRef field. `Type` should be set to
+	//   "ImplementationSpecific" for custom filters.
+	//
+	// Implementers are encouraged to define custom implementation
+	// types to extend the core API with implementation-specific behavior.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
 
 	// ExtensionRef is an optional, implementation-specific extension to the
 	// "filter" behavior.  The resource may be "configmap" (use the empty
 	// string for the group) or an implementation-defined resource (for
 	// example, resource "myroutefilters" in group "networking.acme.io").
 	// Omitting or specifying the empty string for both the resource and
-	// group indicates that the resource is "configmaps".  If the referent
-	// cannot be found, the "InvalidRoutes" status condition on any Gateway
-	// that includes the HTTPRoute will be true.
-	//
-	// Support: custom
-	//
+	// group indicates that the resource is "configmaps".
+	// ExtensionRef MUST NOT be used for core and extended filters.
 	// +optional
 	ExtensionRef *RouteFilterExtensionObjectReference `json:"extensionRef" protobuf:"bytes,2,opt,name=extensionRef"`
+
+	// Filter-specific configuration definitions for core and extended filters
+
+	RequestHeader *HTTPRequestHeaderConfig `json:"requestHeader" protobuf:"bytes,3,opt,name=requestHeader"`
```

Comment:
Nit: It feels like ExtensionRef should be at the bottom of this HTTPRouteFilter type to match where it appears in other types. Or more specifically, I think RequestHeader should appear before ExtensionRef.

Reply:
Let's pick that up in a follow on PR

Reply:
@robscott The reason for defining extensionRef first is to ensure that we can add filters in a backwards compatible for protobuf field ordering manner in future. Should I move the ExtensionRef at the bottom and keep the ordering same as in the current mainline branch?

Reply:
👍 got it, that makes sense. Hopefully we can remove our protobuf gen soonish and revisit this then.